### ISSUE TITLE
ReJit on attach issues

### ISSUE THREAD

**Issue by davmason** [dotnet/coreclr#27781]
Copied from https://github.com/dotnet/docs/issues/15376#issuecomment-552030971

> 
> 
> Hi @davmason
> I see in another post/issue the attach support is resolved for .NET Core. However we are facing technical challenge. Can you please guide us on the following?
> 
> ReJIT after attach reports a success status but does not trigger a new JIT event for any methods. The new ReJIT API has the same issue.
> ngen modules are already loaded which prevents some native methods from producing JIT events.
> 
> [dotnet/coreclr@b314ed8](https://github.com/dotnet/coreclr/commit/b314ed86b11756834b0936e56c3d841e7407f81d)

**Comment by davmason** [comment]
Hi @vashisthaa,

There are two things likely going on here. First is when you trigger a rejit, the compilation doesn't happen immediately. What happens is the runtime flags the method as needing a rejit and then the compilation is triggered next time the method is called. You will get a callback on `GetReJITParameters` to provide the new IL.

The second thing is that rejitted methods don't provide the same jit callbacks, instead of `JITCompilationStarted` you will see `ReJITCompilationStarted`.

See https://github.com/dotnet/coreclr/blob/master/Documentation/Profiling/davbr-blog-archive/ReJIT%20-%20The%20Basics.md and https://github.com/dotnet/coreclr/blob/master/Documentation/Profiling/ReJIT%20on%20Attach.md for an overview.

**Comment by vashisthaa** [comment]
Hi @davmason ,

We verified it ( .NET Core 3.0 Preview 5 on wards on Windows ) works for some of the methods and not for some of the methods. 
This API ICorProfilerInfo3:: EnumJITedFunctions is not giving us all the methods which  were otherwise JITed when we load the profiler with environment variable obviously the profiler forces not to do inlining and load profile images.

**Comment by davmason** [comment]
> 
> 
> Hi @davmason ,
> 
> We verified it ( .NET Core 3.0 Preview 5 on wards on Windows ) works for some of the methods and not for some of the methods.
> This API ICorProfilerInfo3:: EnumJITedFunctions is not giving us all the methods which were otherwise JITed when we load the profiler with environment variable obviously the profiler forces not to do inlining and load profile images.

Can you give me more information about what exactly you are doing? Are you trying to rejit after profiler attach or are you trying to rejit with a startup profiler?


Can you also tell me what startup flags you are using for your profiler and why you are calling EnumJITtedFunctions?

**Comment by vashisthaa** [comment]
Question: Are you trying to rejit after profiler attach or are you trying to rejit with a startup profiler?
Yes. We are trying to ReJIT after the profiler is attached.

When the profiler attaches, we want to comute the already JITed functions and issue a ReJIT on interesting ones. 

For example, when  ICorProfilerInfo3:: EnumJITedFunctions does not return  Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeAsync when this method was already exercised by the time the profiler is attached.

**Comment by davmason** [comment]
How are you trying to prevent the runtime from loading crossgen images? If you are attaching, then module load would have already occurred and if Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeAsync is in a crossgenned module, it wouldn't show up in EnumJITedFunctions.

**Comment by vashisthaa** [comment]
Exactly that is my worry since the Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeAsync is in crossgenned module, it is not possible to attach and issue a ReJIT, instead it is completely possible to get it JITCompile when the CLR profiler is loaded along with the .NET runtime.

Is there any work around?

**Comment by davmason** [comment]
As a work around, instead of iterating through the jitted methods you can 

1. find the module with ICorProfilerInfo3::EnumModules
2. get the metadata interface for that module with ICorProfilerInfo::GetModuleMetadata
3. find the method token through the metadata interface
4. get the FunctionID with ICorProfilerInfo2::GetFunctionFromTokenAndTypeArgs

GetFunctionFromTokenAndTypeArgs is a little dangerous because you can cause things to be loaded that aren't already loaded, and if that happens it can cause deadlocks if you're already in the middle of loading something.

Depending on your scenario, you can also use GetFunctionFromIP3 if you are walking the stack and the function you care about is on the stack. I don't think it will be helpful in your scenario but thought I'd mention it.

**Comment by vashisthaa** [comment]
Thanks for the quick answer and suggestion.
Let's say if I am able to get the function id, will it still be possible to request ReJIT for the methods which are part of crossgenned module?

**Comment by davmason** [comment]
> Thanks for the quick answer and suggestion.
> Let's say if I am able to get the function id, will it still be possible to request ReJIT for the methods which are part of crossgenned module?

Yes, you can rejit methods in a crossgenned module. You can’t rewrite IL via SetILFunctionBody but you can ReJIT.

**Comment by vashisthaa** [comment]
>>>You can’t rewrite IL via SetILFunctionBody but you can ReJIT.

Does it mean the IL-Rewriting profiler can not provide a new function body?

**Comment by davmason** [comment]
> > > > You can’t rewrite IL via SetILFunctionBody but you can ReJIT.
> 
> Does it mean the IL-Rewriting profiler can not provide a new function body?

Sorry, I was not specific enough. You can’t rewrite crossgenned functions at module load time with ICorProfilerInfo::SetILFunctionBody, but you can rejit them and supply new IL via GetReJITParameters by calling ICorProfilerFunctionControl::SetILFunctionBody.

**Comment by vashisthaa** [comment]
Thanks! We were successful in achieving the ReJIT by following the below API steps: Do you foresee any issues with this ?

Step 1) ICorProfilerInfo3::EnumModules
Step 2) ICorProfilerInfo::GetModuleMetaData ( to get the IMetaDataImport for the module)
Step 3) IMetaDataImport::FindTypeDefByName
Step 4) IMetaDataImport::FindMethod
Step 5) ICorProfilerInfo4.RequestReJIT  ( for given module id  and method def )

Thanks

**Comment by davmason** [comment]
> 
> 
> Thanks! We were successful in achieving the ReJIT ( for crossgenned module method) by following the below API steps: Do you foresee any issues with this ?
> 
> Step 1) ICorProfilerInfo3::EnumModules
> Step 2) ICorProfilerInfo::GetModuleMetaData ( to get the IMetaDataImport for the module)
> Step 3) IMetaDataImport::FindTypeDefByName
> Step 4) IMetaDataImport::FindMethod
> Step 5) ICorProfilerInfo4.RequestReJIT ( for given module id and method def )
> 
> Thanks

The only potential issue is what I mentioned above that `GetFunctionFromTokenAndTypeArgs` can force things to be loaded if they aren't already, and if you force the wrong thing to be loaded it can deadlock.
